File: src/dnfw/asciiglitch.py
"""A mark decomposed into glitching ASCII, as frames for the Digitone II's 128 x 64 screen.

The boot-screen mod's second animation. It follows DNX's PHOSPHOR loader
(`DNX/web/src/landing/dnxloader.ts`): a grid of character cells starts as
scrambled glyphs, each cell locks into the picture at its own moment (random,
with a slight left-to-right bias), rows tear sideways while the chaos decays,
thin slices slip like broken video sync, and a residual glitch lingers once the
picture has resolved.

The panel is one bit and 128 x 64, so what DNX does with tone, alpha, bloom and
scanlines is folded into which pixels light: a glyph's alpha becomes the chance
it is drawn at all. Cells are 4 x 6 pixels -- a 3 x 5 glyph of our own font
(`font3x5.json`) and a pixel of gap -- so the grid is 32 x 10. Only glyphs in
that font can be used, which is why the character lists are smaller than DNX's.

## Parity

`site/js/asciiglitch.js` is the same algorithm, line for line, and must produce
the same pixels: `scripts/js_asciiglitch_check.mjs` fails otherwise. Both use
only integer hashing and correctly rounded arithmetic -- `u * sqrt(u)` stands in
for DNX's `u ** 1.45`, because `pow` is not correctly rounded and could differ
between Python and a browser in the last bit.
"""

from __future__ import annotations

import json
import math
import pathlib

W, H = 128, 64
CW, CH = 4, 6                 # a cell: a 3 x 5 glyph and a pixel of gap
COLS, ROWS = W // CW, H // CH  # 32 x 10
Y0 = (H - ROWS * CH) // 2      # 2: the grid centred vertically

RAMP = " .:+#"                 # the picture, dark to bright
GLITCH = "01/:=+<>[]{}#%*?_|-\\"

FONT = json.loads(pathlib.Path(__file__).with_name("font3x5.json").read_text())["glyphs"]
M32 = 0xFFFFFFFF


class GlitchError(ValueError):
    """The options cannot make an animation."""


def _imul(a: int, b: int) -> int:
    return ((a & M32) * (b & M32)) & M32


def rnd(key: int) -> float:
    """DNX's stable integer noise, 0 <= r < 1."""
    n = key & M32
    n = _imul(n ^ (n >> 16), 0x45D9F3B)
    n = _imul(n ^ (n >> 16), 0x45D9F3B)
    return ((n ^ (n >> 16)) & M32) / 4294967296


def _smooth(t: float) -> float:
    t = min(1.0, max(0.0, t))
    return t * t * (3 - 2 * t)


def check_chars(chars: str, what: str) -> str:
    chars = chars.upper()
    missing = sorted({c for c in chars if c not in FONT})
    if missing:
        raise GlitchError(f"{what}: no glyph for {''.join(missing)!r}; the font has {''.join(FONT)!r}")
    if not chars:
        raise GlitchError(f"{what}: give at least one character")
    return chars


def cells(lit) -> list[str]:
    """The picture as characters: each cell's lit coverage picks from `ramp`.
    Returns the coverage counts (0..24) per cell, row by row."""
    out = []
    for r in range(ROWS):
        for c in range(COLS):
            n = 0
            for y in range(Y0 + r * CH, Y0 + r * CH + CH):
                for x in range(c * CW, c * CW + CW):
                    n += 1 if lit(x, y) else 0
            out.append(n)
    return out


def _draw(px: bytearray, ch: str, x0: int, y0: int) -> None:
    for gy, row in enumerate(FONT[ch]):
        y = y0 + gy
        if y < 0 or y >= H:
            continue
        for gx, bit in enumerate(row):
            x = x0 + gx
            if bit == "1" and 0 <= x < W:
                px[y * W + x] = 1


def frames(lit, *, ramp: str = RAMP, glitch_chars: str = GLITCH, resolve: int = 40,
           idle_frames: int = 16, seed: int = 26, glitch: float = 1.0,
           idle: float = 0.3) -> list[bytearray]:
    """-> `resolve + idle_frames` frames, each 128 x 64 bytes of 0/1, row-major.

    `lit(x, y)` is the picture. Frames `0..resolve-1` go from noise to the
    picture; the rest are the settled picture with residual glitches, meant to
    loop. `glitch` scales the chaos (0 = no tearing or scramble, 1 = DNX's),
    `idle` the residual glitch."""
    ramp = check_chars(ramp, "picture characters")
    glitch_chars = check_chars(glitch_chars, "glitch characters")
    if resolve < 1 or idle_frames < 0:
        raise GlitchError("resolve needs at least one frame, and idle frames cannot be negative")
    if not 0 <= glitch <= 2 or not 0 <= idle <= 1:
        raise GlitchError("glitch is 0..2 and idle 0..1")
    cover = cells(lit)
    steps = len(ramp) - 1
    target = [ramp[(n * steps * 2 + 24) // 48] for n in cover]
    g = len(glitch_chars)
    grid = []
    for r in range(ROWS):
        for c in range(COLS):
            key = (c + 7) * 733 + (r + 4) * 7919 + seed
            grid.append((key, 0.035 + rnd(key + 19) * 0.84 + (c / COLS) * 0.11))

    out = []
    for k in range(resolve + idle_frames):
        progress = (k / (resolve - 1) if resolve > 1 else 1.0) if k < resolve else 1.0
        reveal = _smooth((progress - 0.10) / 0.90)
        u = 1 - reveal
        chaos = u * math.sqrt(u) * glitch
        idle_amt = idle if k >= resolve else idle * _smooth((progress - 0.72) / 0.28)
        row_tick = k >> 1
        px = bytearray(W * H)
        for r in range(ROWS):
            row_key = r * 919 + row_tick * 311 + seed
            tearing = 2.4 if rnd(row_key + 41) < 0.22 else 1.0
            sx = math.floor((rnd(row_key) - 0.5) * CW * 15 * chaos * tearing + 0.5)
            sy = math.floor((rnd(row_key + 11) - 0.5) * CH * 1.8 * chaos * chaos + 0.5)
            if k >= resolve:
                j = k - resolve
                if rnd(seed * 31 + j * 977 + r * 13) < idle * 0.08:
                    step = 1 + math.floor(rnd(seed + j * 71 + r * 7) * 3)
                    sx += step if rnd(seed + j * 53 + r) < 0.5 else -step
            for c in range(COLS):
                key, lock_at = grid[r * COLS + c]
                if reveal >= lock_at:
                    ch = target[r * COLS + c]
                    if ch != " " and rnd(key + k * 1103) < idle_amt * 0.04:
                        ch = glitch_chars[math.floor(rnd(key + k * 97) * g)]
                    if ch == " ":
                        continue
                else:
                    n = rnd(key + k * 13007)
                    if n > 0.50 + chaos * 0.12:
                        continue
                    ch = glitch_chars[math.floor(rnd(key + k * 701 + 7) * g)]
                    alpha = 0.24 + rnd(key + k * 503) * 0.72
                    if rnd(key + k * 211) > alpha:
                        continue
                _draw(px, ch, c * CW + sx, Y0 + r * CH + sy)
        if chaos > 0.04:
            for i in range(4):
                yy = math.floor(rnd(row_tick * 263 + i * 73 + seed) * H)
                hh = max(1, math.floor(CH * (0.08 + chaos * 0.5) + 0.5))
                dx = math.floor((rnd(i * 31 + row_tick * 911) - 0.5) * CW * chaos * 18 + 0.5)
                for y in range(yy, min(H, yy + hh)):
                    row = px[y * W:(y + 1) * W]
                    for x in range(W):
                        s = x - dx
                        px[y * W + x] = row[s] if 0 <= s < W else 0
        out.append(px)
    return out

File: site/js/asciiglitch.js
/**
 * A mark decomposed into glitching ASCII, as frames for the 128 x 64 screen.
 *
 * The browser half of `src/dnfw/asciiglitch.py`, which explains the design (a
 * port of DNX's PHOSPHOR loader to one bit and a 3 x 5 font). Same names, same
 * order of operations, same arithmetic -- `scripts/js_asciiglitch_check.mjs`
 * requires identical pixels.
 */

import { FONT } from "./font3x5.js";

export const W = 128, H = 64, CW = 4, CH = 6;
export const COLS = W / CW, ROWS = Math.floor(H / CH);
const Y0 = Math.floor((H - ROWS * CH) / 2);
export const RAMP = " .:+#";
export const GLITCH = "01/:=+<>[]{}#%*?_|-\\";
export const GLYPHS = Object.keys(FONT).join("");

export class GlitchError extends Error {}

/** DNX's stable integer noise, 0 <= r < 1. */
export function rnd(key) {
  let n = key >>> 0;
  n = Math.imul(n ^ (n >>> 16), 0x45d9f3b) >>> 0;
  n = Math.imul(n ^ (n >>> 16), 0x45d9f3b) >>> 0;
  return ((n ^ (n >>> 16)) >>> 0) / 4294967296;
}

const smooth = (t) => { t = Math.min(1, Math.max(0, t)); return t * t * (3 - 2 * t); };

export function checkChars(chars, what) {
  chars = String(chars).toUpperCase();
  const missing = [...new Set([...chars].filter((c) => !(c in FONT)))].sort();
  if (missing.length) throw new GlitchError(`${what}: no glyph for "${missing.join("")}"`);
  if (!chars.length) throw new GlitchError(`${what}: give at least one character`);
  return chars;
}

export function cells(lit) {
  const out = [];
  for (let r = 0; r < ROWS; r++) {
    for (let c = 0; c < COLS; c++) {
      let n = 0;
      for (let y = Y0 + r * CH; y < Y0 + r * CH + CH; y++) {
        for (let x = c * CW; x < c * CW + CW; x++) n += lit(x, y) ? 1 : 0;
      }
      out.push(n);
    }
  }
  return out;
}

function draw(px, ch, x0, y0) {
  FONT[ch].forEach((row, gy) => {
    const y = y0 + gy;
    if (y < 0 || y >= H) return;
    for (let gx = 0; gx < row.length; gx++) {
      const x = x0 + gx;
      if (row[gx] === "1" && x >= 0 && x < W) px[y * W + x] = 1;
    }
  });
}

/**
 * -> `resolve + idleFrames` frames, each a 128 x 64 Uint8Array of 0/1, row-major.
 * Frames 0..resolve-1 go from noise to the picture; the rest loop.
 */
export function frames(lit, {
  ramp = RAMP, glitchChars = GLITCH, resolve = 40, idleFrames = 16,
  seed = 26, glitch = 1.0, idle = 0.3,
} = {}) {
  ramp = checkChars(ramp, "picture characters");
  glitchChars = checkChars(glitchChars, "glitch characters");
  if (!(resolve >= 1) || !(idleFrames >= 0)) {
    throw new GlitchError("resolve needs at least one frame, and idle frames cannot be negative");
  }
  if (!(glitch >= 0 && glitch <= 2) || !(idle >= 0 && idle <= 1)) throw new GlitchError("glitch is 0..2 and idle 0..1");
  const cover = cells(lit);
  const steps = ramp.length - 1;
  const target = cover.map((n) => ramp[Math.floor((n * steps * 2 + 24) / 48)]);
  const g = glitchChars.length;
  const grid = [];
  for (let r = 0; r < ROWS; r++) {
    for (let c = 0; c < COLS; c++) {
      const key = (c + 7) * 733 + (r + 4) * 7919 + seed;
      grid.push([key, 0.035 + rnd(key + 19) * 0.84 + (c / COLS) * 0.11]);
    }
  }

  const out = [];
  for (let k = 0; k < resolve + idleFrames; k++) {
    const progress = k < resolve ? (resolve > 1 ? k / (resolve - 1) : 1.0) : 1.0;
    const reveal = smooth((progress - 0.10) / 0.90);
    const u = 1 - reveal;
    const chaos = u * Math.sqrt(u) * glitch;
    const idleAmt = k >= resolve ? idle : idle * smooth((progress - 0.72) / 0.28);
    const rowTick = k >> 1;
    const px = new Uint8Array(W * H);
    for (let r = 0; r < ROWS; r++) {
      const rowKey = r * 919 + rowTick * 311 + seed;
      const tearing = rnd(rowKey + 41) < 0.22 ? 2.4 : 1.0;
      let sx = Math.floor((rnd(rowKey) - 0.5) * CW * 15 * chaos * tearing + 0.5);
      const sy = Math.floor((rnd(rowKey + 11) - 0.5) * CH * 1.8 * chaos * chaos + 0.5);
      if (k >= resolve) {
        const j = k - resolve;
        if (rnd(seed * 31 + j * 977 + r * 13) < idle * 0.08) {
          const step = 1 + Math.floor(rnd(seed + j * 71 + r * 7) * 3);
          sx += rnd(seed + j * 53 + r) < 0.5 ? step : -step;
        }
      }
      for (let c = 0; c < COLS; c++) {
        const [key, lockAt] = grid[r * COLS + c];
        let ch;
        if (reveal >= lockAt) {
          ch = target[r * COLS + c];
          if (ch !== " " && rnd(key + k * 1103) < idleAmt * 0.04) {
            ch = glitchChars[Math.floor(rnd(key + k * 97) * g)];
          }
          if (ch === " ") continue;
        } else {
          const n = rnd(key + k * 13007);
          if (n > 0.50 + chaos * 0.12) continue;
          ch = glitchChars[Math.floor(rnd(key + k * 701 + 7) * g)];
          const alpha = 0.24 + rnd(key + k * 503) * 0.72;
          if (rnd(key + k * 211) > alpha) continue;
        }
        draw(px, ch, c * CW + sx, Y0 + r * CH + sy);
      }
    }
    if (chaos > 0.04) {
      for (let i = 0; i < 4; i++) {
        const yy = Math.floor(rnd(rowTick * 263 + i * 73 + seed) * H);
        const hh = Math.max(1, Math.floor(CH * (0.08 + chaos * 0.5) + 0.5));
        const dx = Math.floor((rnd(i * 31 + rowTick * 911) - 0.5) * CW * chaos * 18 + 0.5);
        for (let y = yy; y < Math.min(H, yy + hh); y++) {
          const row = px.slice(y * W, (y + 1) * W);
          for (let x = 0; x < W; x++) {
            const s = x - dx;
            px[y * W + x] = s >= 0 && s < W ? row[s] : 0;
          }
        }
      }
    }
    out.push(px);
  }
  return out;
}
